### PR TITLE
Allow a request object by considering a clock skew for smooth interoperability

### LINKED ISSUE #34804: Allow a request object by considering a clock skew for smooth interoperability
### Description

By follwing [the latest FAPI 2.0 security profiles draft](https://openid.bitbucket.io/fapi/fapi-2_0-security-profile.html#name-general-requirements), 

An authorization server should allow the following request object: 
- whose "iat" and "nbf" claim is within 0 to 10 seconds ahead in the future.

An authorization server should reject the follwoing request object:
- whose "iat" and "nbf" claim is over 60 seconds ahead in the future.

Whether a request object whose "iat" and "nbf" claim is within 11 to 60 seconds ahead in the future is accepted or not is up to the authorization server.

The aim of this enhancement is to achieve smooth interoperability. The specification describes this point.

I think the following two options:
1. Add a new configuration to determine a clock skew for JWT that a client crafts: a request object, client assertion, DPoP Proof
2. Add a hard-coded parameter to determine the clock skew - 15 seconds.

I would like to take the option 2 because adding a new configuration add some maintainance costs for Keycloak users and developer.

### Discussion

_No response_

### Motivation

To make Keycloak comply with FAPI 2.0 seccurity profiles, the enhancement is needed.
By using OAuth-SIG's automated testing platform, Keycloak lacks the enhancement.
https://github.com/keycloak/kc-sig-fapi/issues/691

### Details

_No response_

### PR DESCRIPTION
**Pull request by tnorimat**
closes #34804

### CODE REVIEW

**Comment by tnorimat** [comment]
@mposolda Hello, if possible, could you review the PR? It is needed to pass the latest draft version of FAPI 2.0 security profiles.

**Comment by thomasdarimont** [comment]
Do you also plan to apply the skew for the `iat` claim?

**Comment by tnorimat** [comment]
@thomasdarimont Hello, according to the draft, 0 to 10 seconds skew is acceptable Over 60 second skew is not acceptable. Whether 10 to 60 skew is acceptable of not up to an authorization server product (described in NOTE 4 of the draft).
There is no specific reason why I choose 15 sec acceptable skew as a defaut setting, but user can change the skew by `SecureRequestObjectExecutor`.

**Comment by tnorimat** [comment]
@thomasdarimont The conformance test for FAPI 2.0 only checks "nbf", not "iat".

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java [hunk 1]
```diff
@@ -560,12 +562,18 @@ public void testSecureRequestObjectExecutor() throws Exception {
...
-        requestObject.nbf(requestObject.getNbf() + 600);
+        requestObject.nbf(requestObject.getNbf() + allowedClockSkew + 2);
         registerRequestObject(requestObject, clientId, Algorithm.ES256, false);
         oauth.openLoginForm();
         assertEquals(OAuthErrorException.INVALID_REQUEST_OBJECT, oauth.getCurrentQuery().get(OAuth2Constants.ERROR));
         assertEquals("Request not yet being processed", oauth.getCurrentQuery().get(OAuth2Constants.ERROR_DESCRIPTION));
 
+        // nbf ahead within allowed clock skew
+        requestObject = createValidRequestObjectForSecureRequestObjectExecutor(clientId);
+        requestObject.nbf(requestObject.getNbf() + allowedClockSkew - 2);
```
Comment (mposolda) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java:573]: Minor: Is it possible to use bigger interval? For example something like ` ... + allowedClockSkew - 10` ? I afraid that if it is only 2 seconds, the test can fail in some slow environments.

**Review by mposolda — COMMENTED** [review]
I've added one minor comment inline regarding testsuite.

@thomasdarimont @tnorimat +1 for verify `iat` as well. I suppose we can use same `allowed-clock-skew` for both `iat` and `nbf`?

Also I think that if `iat` is present on the request object, we can always validate it. So maybe we can go without any additional config option for the `iat` and just have the config option `allowed-clock-skew` introduced in this PR?

**Comment by tnorimat** [comment]
@thomasdarimont @mposolda Thank you for the comment. I will also verify `iat` with clock skew as a default.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java (hunk 1 above)
Comment (tnorimat) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/client/policies/ClientPoliciesExecutorTest.java:573]: Yes, I will use the bigger interval as you suggested.

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks!

**Pull request merged**